Add experimental batched COO sparse format

jakevdp opened the pull request:
This is an implementation of a batch-friendly multidimensional COO format for JAX. It contains implementations of four primitives:

- `bcoo_todense`
- `bcoo_fromdense`
- `bcoo_extract`
- `bcoo_dot_general`

As well as batching, JVP, and transpose rules for each.

For convenience, this also adds class `BCOO`, which is a pytree wrapper around these.

jakevdp commented:
@froystig This is ready for a review.

froystig reviewed (approved):
whew, `dot_general`!

jakevdp committed:
Add experimental batched COO sparse format.

This is an implementation of a batch-friendly multidimensional COO format for JAX. It contains implementations of four primitives (bcoo_todense, bcoo_fromdense, bcoo_extract, bcoo_dot_general), as well as batching, JVP, and transpose rules for each.

For convenience, this also adds class BCOO, which is a pytree wrapper around these.

jakevdp commented on jax/experimental/sparse_ops.py:
```diff
@@ -48,6 +49,9 @@
 from jax.interpreters import ad
 from jax import ad_util
+from jax._src.lax import lax
+from jax._src.util import safe_zip
```
Turns out I was using some private utilities... I changed to importing these explicitly.
